src/metabase/query_processor/pivot/postprocess.clj

```diff
@@ -260,62 +260,79 @@
-  (concat
+  (perf/concat
    (if (= 1 (count pivot-measures))
-     (mapv (fn [col-combo] (mapv fmt col-formatters col-combo)) col-combos)
-     (for [col-combo col-combos
-           measure-key pivot-measures]
-       (conj
-        (mapv fmt col-formatters col-combo)
-        (get column-titles measure-key))))
+     (mapv (fn [col-combo] (perf/mapv fmt col-formatters col-combo)) col-combos)
+     (into [] (mapcat (fn [col-combo]
+                        (let [formatted (perf/mapv fmt col-formatters col-combo)
+                              it (.iterator ^Iterable pivot-measures)]
+                          (loop [acc (transient [])]
+                            (if (.hasNext it)
+                              (recur (conj! acc (conj formatted (get column-titles (.next it)))))
+                              (persistent! acc))))))
+           col-combos))
    (repeat (count pivot-measures)
-           (concat
+           (perf/concat
             (when (and row-totals? (> (count pivot-cols) 0)) ["Row totals"])
             (repeat (dec (count pivot-cols)) nil)
             (when (and (seq pivot-cols) (> (count pivot-measures) 1)) [nil])))))
 
 (defn- build-headers
   "Combine row keys with column headers."
   [column-headers {:keys [pivot-cols pivot-rows column-titles]}]
-  (some->> (not-empty (filter seq column-headers))
-           (apply mapv vector)
+  (some->> (not-empty (filterv seq column-headers))
+           perf/transpose
            (mapv (fn [h]
-                   (-> (mapv #(get column-titles %)
-                             (if (and (seq pivot-cols) (empty? pivot-rows))
-                               pivot-cols pivot-rows))
-                       (into h))))))
+                   (perf/concat
+                    (perf/mapv #(get column-titles %)
+                               (if (and (seq pivot-cols) (empty? pivot-rows))
+                                 pivot-cols pivot-rows))
+                    h)))))
 
 (defn- build-row
   "Build a single row of the pivot table."
   [row-combo col-combos pivot-measures data totals row-totals? ordered-formatters row-formatters config]
+  ;; This implementation is very unorthodox, but this function is incredibly hot, so it must avoid allocation at all
+  ;; readiability costs. Any iterator-based iteration allocates, but also constructing an internal mapping lambda
+  ;; that encloses the outer value is so so expensive.
   (let [row-path       (vec row-combo)
         row-data       (get-in data row-path)
-        ;; Mutable ArrayList intentional because of the hotness of the function.
-        measure-values (ArrayList. (* (max 1 (count col-combos)) (count pivot-measures)))]
+        n (count col-combos)
+        m (count pivot-measures)
+        result (ArrayList. (* (max 1 n) m))]
+    (when-not (seq row-formatters)
+      (dotimes [_ (count pivot-measures)] (.add result nil)))
+    (perf/run! #(.add result %) row-combo)
```

Comment:
nit - the reason for this wasn't immediately obvious to me, maybe worth a comment?

Reply:
I'll try to come up with something.